{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 155,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Assumes one model to one sample and one sample to one passage\n",
      "also assumes both sample platform and Omic will have model ID's on every row\n"
     ]
    }
   ],
   "source": [
    "print(\"Assumes one model to one sample and one sample to one passage\")\n",
    "print(\"also assumes both sample platform and Omic will have model ID's on every row\")\n",
    "providerAbbr = \"PMLB\"\n",
    "os.chdir(\"/home/afollette/jupyter_notebooks/populateOmicFromSamplePlatform/\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 156,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "sending incremental file list\n",
      "cna/\n",
      "cna/PMLB_cna.tsv\n",
      "mut/\n",
      "mut/PMLB_mut.tsv\n",
      "\n",
      "sent 785,375 bytes  received 66 bytes  1,570,882.00 bytes/sec\n",
      "total size is 31,935,310  speedup is 40.66\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "mkdir: cannot create directory ‘workingFiles’: File exists\n"
     ]
    }
   ],
   "source": [
    "%%bash -s \"$providerAbbr\"\n",
    "mkdir workingFiles \n",
    "rsync -avz  /home/afollette/Finder_Data_Repositories/pdxfinder-data/data/UPDOG/\"$1\"/* ./workingFiles"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 157,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as ps\n",
    "import os\n",
    "import glob\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 158,
   "metadata": {},
   "outputs": [],
   "source": [
    "def useSamplePlatformToPopulateOmic(omicDf,omicType):\n",
    "    print(\"Running dataframe for datatype {}\".format(omicType))\n",
    "    columnsList = omicDf.columns.tolist()\n",
    "    populatedOmic = ps.DataFrame()\n",
    "    tmpDf = ps.DataFrame()\n",
    "    for index, row in omicDf.iterrows():\n",
    "        spRows = spDf[(spDf['sample_id'] == str(row['sample_id'])) & (spDf['molecular_characterisation_type'] == omicType)]           \n",
    "        if len(spRows) == 1 :\n",
    "            spRow = spRows.iloc[0]\n",
    "            row['model_id']  = spRow['model_id']\n",
    "            row['sample_origin'] = spRow['sample_origin']\n",
    "            row['host_strain_nomenclature'] = spRow['host_strain_nomenclature']\n",
    "            row['passage'] = spRow['passage']\n",
    "            row['platform'] = spRow['platform']\n",
    "            tmpDf = populatedOmic.append(row)\n",
    "            populatedOmic = tmpDf\n",
    "        else :\n",
    "            print(\"No match for sample id {}\".format(row['sample_id']))\n",
    "    newHeaders = populatedOmic[columnsList]\n",
    "    newFileName = \"PMLB_pop_{0}.tsv\".format(omicType)\n",
    "    newHeaders.to_csv(newFileName, sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 159,
   "metadata": {},
   "outputs": [],
   "source": [
    "def samplePlatformIsAmbiguous(spDf):\n",
    "    isAmbiguous = True\n",
    "    duplicatedSampleId = spDf.loc[ : , ['sample_id', 'molecular_characterisation_type']].duplicated()\n",
    "    duplicatedModels = spDf[duplicatedSampleId].loc[:, ['model_id']].duplicated()\n",
    "    if len(spDf[duplicatedSampleId].values) != 0:\n",
    "        print(\"ambiguous sample id's detected. Aborting process\")\n",
    "    elif len(duplicatedModels.values) != 0:\n",
    "        print(\"amibguious models detected. Aborting process\")\n",
    "    else:\n",
    "        isAmbiguous = False\n",
    "    return isAmbiguous"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 160,
   "metadata": {},
   "outputs": [],
   "source": [
    "try:\n",
    "    os.chdir('./workingFiles')\n",
    "except:\n",
    "    print('Already in ./workingFiles')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 161,
   "metadata": {},
   "outputs": [],
   "source": [
    "mutData = glob.glob(\"./mut/*_mut.tsv\")[0]\n",
    "cnaData = glob.glob(\"./cna/*_cna.tsv\")[0]\n",
    "spData = glob.glob(\"./*_sampleplatform*.tsv\")[0]\n",
    "spDf = ps.read_csv(spData, sep='\\t', dtype=str)\n",
    "spMut = spDf[spDf.molecular_characterisation_type == \"mutation\"]\n",
    "spCNA = spDf[spDf.molecular_characterisation_type == \"copy number alteration\"]\n",
    "mutDf = ps.read_csv(mutData,sep='\\t', dtype=str)\n",
    "cnaDf = ps.read_csv(cnaData, sep='\\t', dtype=str)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 162,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Passed samplePlatform checks. Starting to populate data\n"
     ]
    }
   ],
   "source": [
    "if not samplePlatformIsAmbiguous(spDf):\n",
    "    print(\"Passed samplePlatform checks. Starting to populate data\")\n",
    "    (spMut\n",
    "            .merge(mutDf, on='sample_id', validate='one_to_many', how='inner')\n",
    "            .reindex(columns=['model_id_x', 'sample_id', 'sample_origin_x', 'host_strain_nomenclature_x','passage_x', \"symbol\",\"biotype\",\"coding_sequence_change\",\"variant_class\",\"codon_change\",\"amino_acid_change\",\"consequence\",\"functional_prediction\",\"read_depth\",\"allele_frequency\",\"chromosome\",\"seq_start_position\",\"ref_allele\",\"alt_allele\",\"ucsc_gene_id\",\"ncbi_gene_id\",\"ncbi_transcript_id\",\"ensembl_gene_id\",\"ensembl_transcript_id\",\"variation_id\",\"genome_assembly\",\"platform\"])\n",
    "            .to_csv(\"./mut/\" + providerAbbr + \"_mut.tsv\", index=False, sep='\\t')        \n",
    "               )\n",
    "    \n",
    "    testFrame =(spMut\n",
    "            .merge(mutDf, on='sample_id', validate='one_to_many', how='left')\n",
    "            .query('model_id_y != model_id_y')\n",
    "             .to_csv(\"./mut/Left_droppedPoints.tsv\",index=False, sep='\\t')\n",
    "               )\n",
    "    \n",
    "    (spMut\n",
    "            .merge(mutDf, on='sample_id', validate='one_to_many', how='right')\n",
    "            .query('model_id_x != model_id_x')\n",
    "            .to_csv(\"./mut/right_droppedPoints.tsv\",index=False, sep='\\t')\n",
    "     )\n",
    "    \n",
    "    \n",
    "    (spCNA\n",
    "            .merge(mutDf, on='sample_id', validate='one_to_many', how='inner')\n",
    "            .reindex(columns=['model_id_x', 'sample_id', 'sample_origin_x', 'passage_x', 'host_strain_nomenclature_x', \"model_id\",\"sample_id\",\"sample_origin\",\"passage\",\"host_strain_nomenclature\",\"chromosome\",\"seq_start_position\",\"seq_end_position\",\"symbol\",\"ucsc_gene_id\",\"ncbi_gene_id\",\"ensembl_gene_id\",\"log10r_cna\",\"log2r_cna\",\"fold_change\",\"copy_number_status\",\"gistic_value\",\"picnic_value\",\"genome_assembly\",\"platform\"])\n",
    "            .to_csv(\"./cna/\" + providerAbbr + \"_cna.tsv\", index=False, sep='\\t')        \n",
    "               )\n",
    "    (spCNA\n",
    "            .merge(mutDf, on='sample_id', validate='one_to_many', how='left')\n",
    "            .query('model_id_y != model_id_y')\n",
    "            .to_csv(\"./cna/left_droppedPoints.tsv\",index=False, sep='\\t')\n",
    "     )\n",
    "    \n",
    "    (spCNA\n",
    "            .merge(mutDf, on='sample_id', validate='one_to_many', how='right')\n",
    "            .query('model_id_x != model_id_x')\n",
    "            .to_csv(\"./cna/right_droppedPoints.tsv\",index=False, sep='\\t')\n",
    "     )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 153,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Field</th>\n",
       "      <th>sample_id</th>\n",
       "      <th>sample_origin_x</th>\n",
       "      <th>passage_x</th>\n",
       "      <th>engrafted_tumor_collection_site</th>\n",
       "      <th>model_id_x</th>\n",
       "      <th>host_strain_name</th>\n",
       "      <th>host_strain_nomenclature_x</th>\n",
       "      <th>molecular_characterisation_type</th>\n",
       "      <th>platform_x</th>\n",
       "      <th>...</th>\n",
       "      <th>ref_allele</th>\n",
       "      <th>alt_allele</th>\n",
       "      <th>ucsc_gene_id</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>ncbi_transcript_id</th>\n",
       "      <th>ensembl_gene_id</th>\n",
       "      <th>ensembl_transcript_id</th>\n",
       "      <th>variation_id</th>\n",
       "      <th>genome_assembly</th>\n",
       "      <th>platform_y</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Example</td>\n",
       "      <td>CRC0228PRH0000000000D01000</td>\n",
       "      <td>xenograft</td>\n",
       "      <td>0</td>\n",
       "      <td>mammary glad</td>\n",
       "      <td>CRC0228PR</td>\n",
       "      <td>NOD SCID</td>\n",
       "      <td>NOD.Cg-Prkdcscid Il2rgtm1Wjl/SzJ</td>\n",
       "      <td>mutation</td>\n",
       "      <td>Affymetrix SNP6.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90</th>\n",
       "      <td>NaN</td>\n",
       "      <td>64842</td>\n",
       "      <td>patient</td>\n",
       "      <td>non applicable</td>\n",
       "      <td>non applicable</td>\n",
       "      <td>PMLB64842</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>mutation</td>\n",
       "      <td>Targeted Next generation sequencing</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>166</th>\n",
       "      <td>NaN</td>\n",
       "      <td>73123</td>\n",
       "      <td>patient</td>\n",
       "      <td>non applicable</td>\n",
       "      <td>non applicable</td>\n",
       "      <td>PMLB73123</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>mutation</td>\n",
       "      <td>Targeted Next generation sequencing</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>480</th>\n",
       "      <td>NaN</td>\n",
       "      <td>2393</td>\n",
       "      <td>xenograft</td>\n",
       "      <td>3</td>\n",
       "      <td>mammary fat pad</td>\n",
       "      <td>PMLB2393</td>\n",
       "      <td>NOD SCID</td>\n",
       "      <td>NOD.CB17-Prkdcscid/J</td>\n",
       "      <td>mutation</td>\n",
       "      <td>Targeted Next generation sequencing</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>481</th>\n",
       "      <td>NaN</td>\n",
       "      <td>2435</td>\n",
       "      <td>xenograft</td>\n",
       "      <td>3</td>\n",
       "      <td>mammary fat pad</td>\n",
       "      <td>PMLB2435</td>\n",
       "      <td>NOD SCID</td>\n",
       "      <td>NOD.CB17-Prkdcscid/J</td>\n",
       "      <td>mutation</td>\n",
       "      <td>Targeted Next generation sequencing</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4277</th>\n",
       "      <td>NaN</td>\n",
       "      <td>POP_176_Li_X</td>\n",
       "      <td>xenograft</td>\n",
       "      <td>2</td>\n",
       "      <td>flank</td>\n",
       "      <td>PMLB_POP176</td>\n",
       "      <td>NOD SCID</td>\n",
       "      <td>NOD.CB17-Prkdcscid/J</td>\n",
       "      <td>mutation</td>\n",
       "      <td>whole exome sequencing</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4278</th>\n",
       "      <td>NaN</td>\n",
       "      <td>POP_177_Li_X</td>\n",
       "      <td>xenograft</td>\n",
       "      <td>1</td>\n",
       "      <td>flank</td>\n",
       "      <td>PMLB_POP177</td>\n",
       "      <td>NOD SCID</td>\n",
       "      <td>NOD.CB17-Prkdcscid/J</td>\n",
       "      <td>mutation</td>\n",
       "      <td>whole exome sequencing</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4279</th>\n",
       "      <td>NaN</td>\n",
       "      <td>POP_179_Li_X</td>\n",
       "      <td>xenograft</td>\n",
       "      <td>1</td>\n",
       "      <td>flank</td>\n",
       "      <td>PMLB_POP179</td>\n",
       "      <td>NOD SCID</td>\n",
       "      <td>NOD.CB17-Prkdcscid/J</td>\n",
       "      <td>mutation</td>\n",
       "      <td>whole exome sequencing</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4280</th>\n",
       "      <td>NaN</td>\n",
       "      <td>POP_184_Li_X</td>\n",
       "      <td>xenograft</td>\n",
       "      <td>unknown</td>\n",
       "      <td>flank</td>\n",
       "      <td>PMLB_POP184</td>\n",
       "      <td>NOD SCID</td>\n",
       "      <td>NOD.CB17-Prkdcscid/J</td>\n",
       "      <td>mutation</td>\n",
       "      <td>whole exome sequencing</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4281</th>\n",
       "      <td>NaN</td>\n",
       "      <td>POP_185_Li_X</td>\n",
       "      <td>xenograft</td>\n",
       "      <td>2</td>\n",
       "      <td>flank</td>\n",
       "      <td>PMLB_POP185</td>\n",
       "      <td>NOD SCID</td>\n",
       "      <td>NOD.CB17-Prkdcscid/J</td>\n",
       "      <td>mutation</td>\n",
       "      <td>whole exome sequencing</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>75 rows × 43 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Field                   sample_id sample_origin_x       passage_x  \\\n",
       "0     Example  CRC0228PRH0000000000D01000       xenograft               0   \n",
       "90        NaN                       64842         patient  non applicable   \n",
       "166       NaN                       73123         patient  non applicable   \n",
       "480       NaN                        2393       xenograft               3   \n",
       "481       NaN                        2435       xenograft               3   \n",
       "...       ...                         ...             ...             ...   \n",
       "4277      NaN                POP_176_Li_X       xenograft               2   \n",
       "4278      NaN                POP_177_Li_X       xenograft               1   \n",
       "4279      NaN                POP_179_Li_X       xenograft               1   \n",
       "4280      NaN                POP_184_Li_X       xenograft         unknown   \n",
       "4281      NaN                POP_185_Li_X       xenograft               2   \n",
       "\n",
       "     engrafted_tumor_collection_site   model_id_x host_strain_name  \\\n",
       "0                       mammary glad    CRC0228PR         NOD SCID   \n",
       "90                    non applicable    PMLB64842              NaN   \n",
       "166                   non applicable    PMLB73123              NaN   \n",
       "480                  mammary fat pad     PMLB2393         NOD SCID   \n",
       "481                  mammary fat pad     PMLB2435         NOD SCID   \n",
       "...                              ...          ...              ...   \n",
       "4277                           flank  PMLB_POP176         NOD SCID   \n",
       "4278                           flank  PMLB_POP177         NOD SCID   \n",
       "4279                           flank  PMLB_POP179         NOD SCID   \n",
       "4280                           flank  PMLB_POP184         NOD SCID   \n",
       "4281                           flank  PMLB_POP185         NOD SCID   \n",
       "\n",
       "             host_strain_nomenclature_x molecular_characterisation_type  \\\n",
       "0     NOD.Cg-Prkdcscid Il2rgtm1Wjl/SzJ                         mutation   \n",
       "90                                  NaN                        mutation   \n",
       "166                                 NaN                        mutation   \n",
       "480                NOD.CB17-Prkdcscid/J                        mutation   \n",
       "481                NOD.CB17-Prkdcscid/J                        mutation   \n",
       "...                                 ...                             ...   \n",
       "4277               NOD.CB17-Prkdcscid/J                        mutation   \n",
       "4278               NOD.CB17-Prkdcscid/J                        mutation   \n",
       "4279               NOD.CB17-Prkdcscid/J                        mutation   \n",
       "4280               NOD.CB17-Prkdcscid/J                        mutation   \n",
       "4281               NOD.CB17-Prkdcscid/J                        mutation   \n",
       "\n",
       "                               platform_x  ... ref_allele alt_allele  \\\n",
       "0                       Affymetrix SNP6.0  ...        NaN        NaN   \n",
       "90    Targeted Next generation sequencing  ...        NaN        NaN   \n",
       "166   Targeted Next generation sequencing  ...        NaN        NaN   \n",
       "480   Targeted Next generation sequencing  ...        NaN        NaN   \n",
       "481   Targeted Next generation sequencing  ...        NaN        NaN   \n",
       "...                                   ...  ...        ...        ...   \n",
       "4277               whole exome sequencing  ...        NaN        NaN   \n",
       "4278               whole exome sequencing  ...        NaN        NaN   \n",
       "4279               whole exome sequencing  ...        NaN        NaN   \n",
       "4280               whole exome sequencing  ...        NaN        NaN   \n",
       "4281               whole exome sequencing  ...        NaN        NaN   \n",
       "\n",
       "     ucsc_gene_id ncbi_gene_id ncbi_transcript_id ensembl_gene_id  \\\n",
       "0             NaN          NaN                NaN             NaN   \n",
       "90            NaN          NaN                NaN             NaN   \n",
       "166           NaN          NaN                NaN             NaN   \n",
       "480           NaN          NaN                NaN             NaN   \n",
       "481           NaN          NaN                NaN             NaN   \n",
       "...           ...          ...                ...             ...   \n",
       "4277          NaN          NaN                NaN             NaN   \n",
       "4278          NaN          NaN                NaN             NaN   \n",
       "4279          NaN          NaN                NaN             NaN   \n",
       "4280          NaN          NaN                NaN             NaN   \n",
       "4281          NaN          NaN                NaN             NaN   \n",
       "\n",
       "     ensembl_transcript_id variation_id genome_assembly platform_y  \n",
       "0                      NaN          NaN             NaN        NaN  \n",
       "90                     NaN          NaN             NaN        NaN  \n",
       "166                    NaN          NaN             NaN        NaN  \n",
       "480                    NaN          NaN             NaN        NaN  \n",
       "481                    NaN          NaN             NaN        NaN  \n",
       "...                    ...          ...             ...        ...  \n",
       "4277                   NaN          NaN             NaN        NaN  \n",
       "4278                   NaN          NaN             NaN        NaN  \n",
       "4279                   NaN          NaN             NaN        NaN  \n",
       "4280                   NaN          NaN             NaN        NaN  \n",
       "4281                   NaN          NaN             NaN        NaN  \n",
       "\n",
       "[75 rows x 43 columns]"
      ]
     },
     "execution_count": 153,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "testFrame.query('model_id_y != model_id_y')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
