{
 "cells": [
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-08-09T12:32:55.831735Z",
     "start_time": "2024-08-09T12:32:55.612882Z"
    }
   },
   "cell_type": "code",
   "source": [
    "import pandas as pd\n",
    "from utils import *\n",
    "\n",
    "NKI_submission = \"/Users/tushar/CancerModels/submission/NKI/IDC\"\n",
    "NKI_dir = \"/Users/tushar/CancerModels/pdxfinder-data/data/UPDOG/NKI\"\n",
    "template = \"/Users/tushar/CancerModels/pdxfinder-data/template/active_templates/\""
   ],
   "id": "760e679cf554e6e2",
   "outputs": [],
   "execution_count": 1
  },
  {
   "metadata": {
    "jupyter": {
     "is_executing": true
    },
    "ExecuteTime": {
     "start_time": "2024-08-09T12:32:56.740210Z"
    }
   },
   "cell_type": "code",
   "source": [
    "panelseq_columns = [\"CHROM\", \"POS\", \"ID\", \"REF\", \"ALT\", \"QUAL\", \"FILTER\", \"INFO\", \"FORMAT\", \"IDC025PanelSeqPDX\", \"IDC026PanelSeqPDX\", \"IDC029PanelSeqPDX\", \"IDC031PanelSeqPDX\", \"IDC032PanelSeqPDX\", \"IDC038PPanelSeqDX\", \"IDC057PanelSeqPDX\", \"IDC062PanelSeqPDX\", \"IDC065PanelSeqPDX\", \"IDC069PanelSeqPDX\", \"IDC072PanelSeqPDX\", \"IDC090BPanelSeqPDX\", \"IDC092PanelSeqPDX\", \"IDC097PanelSeqPDX\", \"IDC099PanelSeqPDX\", \"IDC107PanelSeqPDX\", \"IDC113PanelSeqPDX\", \"IDC117PanelSeqPDX\", \"IDC143PanelSeqPDX\", \"IDC152PanelSeqPDX\", \"IDC159APanelSeqPDX\", \"IDC159BPanelSeqPDX\", \"IDC180PanelSeqPDX\", \"IDC186PanelSeqPDX\", \"IDC192PanelSeqPDX\", \"IDC197PanelSeqPDX\", \"IDC198PanelSeqPDX\", \"IDC207PanelSeqPDX\", \"IDC209PanelSeqPDX\", \"IDC216PanelSeqPDX\", \"IDC218PanelSeqPDX\", \"IDC222PanelSeqPDX\", \"IDC223BPanelSeqPDX\", \"IDC229PanelSeqPDX\", \"IDC232PanelSeqPDX\", \"IDC274PanelSeqPDX\", \"IDC282PanelSeqPDX\", \"IDC290PanelSeqPDX\", \"IDC299PanelSeqPDX\", \"IDC307PanelSeqPDX\", \"IDC338PanelSeqPDX\", \"IDC344PanelSeqPDX\", \"IDC346PanelSeqPDX\", \"ILC006PanelSeqPDX\", \"ILC012PanelSeqPDX\", \"ILC083PanelSeqPDX\", \"ILC248PanelSeqPDX\"]\n",
    "mutation_raw_panel_seq = pd.read_csv(join(NKI_submission, 'cm_idc_nki_panelseq_newnames.vcf'),sep='\\t', comment='#',low_memory=False, names=panelseq_columns, header=None)"
   ],
   "id": "b56270c119bcb948",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {
    "jupyter": {
     "is_executing": true
    }
   },
   "cell_type": "code",
   "source": [
    "mutation_raw_panel_seq = mutation_raw_panel_seq.replace('./.:.:.:.:.:.:.:.:.:.', '').replace('./.:.:.:.:.:.:.', '')\n",
    "mutation_raw_panel_seq"
   ],
   "id": "343d6af8aa3dcf47",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {
    "jupyter": {
     "is_executing": true
    }
   },
   "cell_type": "code",
   "source": [
    "mutation_melted = mutation_raw_panel_seq[['CHROM', 'POS', 'ID', 'REF', 'ALT', \"IDC025PanelSeqPDX\", \"IDC026PanelSeqPDX\", \"IDC029PanelSeqPDX\", \"IDC031PanelSeqPDX\", \"IDC032PanelSeqPDX\", \"IDC038PPanelSeqDX\", \"IDC057PanelSeqPDX\", \"IDC062PanelSeqPDX\", \"IDC065PanelSeqPDX\", \"IDC069PanelSeqPDX\", \"IDC072PanelSeqPDX\", \"IDC090BPanelSeqPDX\", \"IDC092PanelSeqPDX\", \"IDC097PanelSeqPDX\", \"IDC099PanelSeqPDX\", \"IDC107PanelSeqPDX\", \"IDC113PanelSeqPDX\", \"IDC117PanelSeqPDX\", \"IDC143PanelSeqPDX\", \"IDC152PanelSeqPDX\", \"IDC159APanelSeqPDX\", \"IDC159BPanelSeqPDX\", \"IDC180PanelSeqPDX\", \"IDC186PanelSeqPDX\", \"IDC192PanelSeqPDX\", \"IDC197PanelSeqPDX\", \"IDC198PanelSeqPDX\", \"IDC207PanelSeqPDX\", \"IDC209PanelSeqPDX\", \"IDC216PanelSeqPDX\", \"IDC218PanelSeqPDX\", \"IDC222PanelSeqPDX\", \"IDC223BPanelSeqPDX\", \"IDC229PanelSeqPDX\", \"IDC232PanelSeqPDX\", \"IDC274PanelSeqPDX\", \"IDC282PanelSeqPDX\", \"IDC290PanelSeqPDX\", \"IDC299PanelSeqPDX\", \"IDC307PanelSeqPDX\", \"IDC338PanelSeqPDX\", \"IDC344PanelSeqPDX\", \"IDC346PanelSeqPDX\", \"ILC006PanelSeqPDX\", \"ILC012PanelSeqPDX\", \"ILC083PanelSeqPDX\", \"ILC248PanelSeqPDX\"]]\n",
    "dcis_columns = [col for col in mutation_melted.columns if 'DCIS' in col]\n",
    "mutation_melted = pd.melt(mutation_melted, id_vars=['CHROM', 'POS', 'ID', 'REF', 'ALT'], value_vars=dcis_columns, var_name='sample_id', value_name='value')\n",
    "mutation_melted = mutation_melted[mutation_melted['value'] != '']\n",
    "column_mapper = {'CHROM': 'chromosome', 'POS': 'seq_start_position', 'REF': 'ref_allele', 'ALT':'alt_allele'}\n",
    "mutation_melted.rename(columns=column_mapper, inplace=True)\n",
    "mutation_template = read_metadata_with_fields(join(template, 'mut/mutation_template_internal-Sheet1.tsv'))\n",
    "mutation_melted['strand'] = '+'\n",
    "mutation_melted['platform_id'] = 'mutation_panel_seq'\n",
    "chromosomes = ['1', '10', '11', '12', '13', '14', '15', '16', '17', '18', '19', '2', '20', '21', '22', '3', '4', '5', '6', '7', '8', '9', 'X', 'Y']\n",
    "mutation_melted = mutation_melted[mutation_melted['chromosome'].isin(chromosomes)]\n",
    "for col in mutation_template.columns:\n",
    "    if col not in mutation_melted:\n",
    "        mutation_melted[col] = ''"
   ],
   "id": "b2ead11b94093168",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {
    "jupyter": {
     "is_executing": true
    }
   },
   "cell_type": "code",
   "source": [
    "submission_mms = read_metadata_with_fields(join(NKI_dir, 'NKI_molecular_metadata-sample.tsv'))\n",
    "submission_model_sample_dict = submission_mms[['model_id', 'sample_id']].drop_duplicates().groupby(['model_id'])['sample_id'].apply(set).to_dict()"
   ],
   "id": "a5aad0a3e92abd1",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": [
    "sample_ids = [\"IDC025PanelSeqPDX\", \"IDC026PanelSeqPDX\", \"IDC029PanelSeqPDX\", \"IDC031PanelSeqPDX\", \"IDC032PanelSeqPDX\", \"IDC038PPanelSeqDX\", \"IDC057PanelSeqPDX\", \"IDC062PanelSeqPDX\", \"IDC065PanelSeqPDX\", \"IDC069PanelSeqPDX\", \"IDC072PanelSeqPDX\", \"IDC090BPanelSeqPDX\", \"IDC092PanelSeqPDX\", \"IDC097PanelSeqPDX\", \"IDC099PanelSeqPDX\", \"IDC107PanelSeqPDX\", \"IDC113PanelSeqPDX\", \"IDC117PanelSeqPDX\", \"IDC143PanelSeqPDX\", \"IDC152PanelSeqPDX\", \"IDC159APanelSeqPDX\", \"IDC159BPanelSeqPDX\", \"IDC180PanelSeqPDX\", \"IDC186PanelSeqPDX\", \"IDC192PanelSeqPDX\", \"IDC197PanelSeqPDX\", \"IDC198PanelSeqPDX\", \"IDC207PanelSeqPDX\", \"IDC209PanelSeqPDX\", \"IDC216PanelSeqPDX\", \"IDC218PanelSeqPDX\", \"IDC222PanelSeqPDX\", \"IDC223BPanelSeqPDX\", \"IDC229PanelSeqPDX\", \"IDC232PanelSeqPDX\", \"IDC274PanelSeqPDX\", \"IDC282PanelSeqPDX\", \"IDC290PanelSeqPDX\", \"IDC299PanelSeqPDX\", \"IDC307PanelSeqPDX\", \"IDC338PanelSeqPDX\", \"IDC344PanelSeqPDX\", \"IDC346PanelSeqPDX\", \"ILC006PanelSeqPDX\", \"ILC012PanelSeqPDX\", \"ILC083PanelSeqPDX\", \"ILC248PanelSeqPDX\"]\n",
    "\n",
    "for mid in submission_model_sample_dict.keys():\n",
    "    temp = mutation_melted[[sid in submission_model_sample_dict[mid] for sid in mutation_melted['sample_id']]]\n",
    "    #model_id = [mid for mid in submission_model_sample_dict if s in submission_model_sample_dict[mid]][0]\n",
    "    #temp = mutation_melted[mutation_melted['sample_id'] == s]\n",
    "    temp[mutation_template.columns].to_csv(join(NKI_submission, f'mut/NKI_mut_{mid}.tsv'), sep='\\t', index=False)"
   ],
   "id": "507ae8c348897758"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": "",
   "id": "8510a4b27998852d"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": "",
   "id": "a99f53679c717be0"
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-08-09T12:18:30.487873Z",
     "start_time": "2024-08-09T12:14:17.975827Z"
    }
   },
   "cell_type": "code",
   "source": "mutation = pd.read_csv(join(NKI_submission, 'cm_idc_nki_panelseq_newnames.vcf'),sep='\\t', comment='#')",
   "id": "2b98a75bc123e19b",
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/_2/g5d3zf4s41g0tvlqkc3_8wjr0000gp/T/ipykernel_93865/3033777406.py:1: DtypeWarning: Columns (0) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  mutation = pd.read_csv(join(NKI_submission, 'cm_idc_nki_panelseq_newnames.vcf'),sep='\\t', comment='#',)\n"
     ]
    }
   ],
   "execution_count": 3
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-08-09T12:18:30.814333Z",
     "start_time": "2024-08-09T12:18:30.572918Z"
    }
   },
   "cell_type": "code",
   "source": "mutation.head()",
   "id": "94a8423245be5769",
   "outputs": [
    {
     "data": {
      "text/plain": [
       "   1  13273  .  G      C ..1                 PASS  \\\n",
       "0  1  13417  .  C  CGAGA   .                 PASS   \n",
       "1  1  13868  .  A      G   .             map_qual   \n",
       "2  1  15118  .  A      G   .             map_qual   \n",
       "3  1  15644  .  G      C   .             map_qual   \n",
       "4  1  15751  .  A      T   .  base_qual;haplotype   \n",
       "\n",
       "  AS_FilterStatus=SITE;AS_SB_TABLE=0,0|0,0;ECNT=1;GERMQ=51;ROQ=54;DP=3;MBQ=0,20;MFRL=0,169;MMQ=60,47;MPOS=15;POPAF=7.3;TLOD=4.64  \\\n",
       "0  AS_FilterStatus=SITE;AS_SB_TABLE=0,0|0,0;ECNT=...                                                                               \n",
       "1  AS_FilterStatus=map_qual;AS_SB_TABLE=0,0|0,0;E...                                                                               \n",
       "2  AS_FilterStatus=map_qual;AS_SB_TABLE=0,0|0,0;E...                                                                               \n",
       "3  AS_FilterStatus=map_qual;AS_SB_TABLE=0,0|0,0;E...                                                                               \n",
       "4  AS_FilterStatus=base_qual;AS_SB_TABLE=0,0|0,0;...                                                                               \n",
       "\n",
       "              GT:AD:AF:DP:F1R2:F2R1:SB                  ./.:.:.:.:.:.:.  ...  \\\n",
       "0             GT:AD:AF:DP:F1R2:F2R1:SB                  ./.:.:.:.:.:.:.  ...   \n",
       "1             GT:AD:AF:DP:F1R2:F2R1:SB                  ./.:.:.:.:.:.:.  ...   \n",
       "2             GT:AD:AF:DP:F1R2:F2R1:SB  0/1:0,2:0.667:2:0,1:0,0:0,0,1,1  ...   \n",
       "3             GT:AD:AF:DP:F1R2:F2R1:SB                  ./.:.:.:.:.:.:.  ...   \n",
       "4  GT:AD:AF:DP:F1R2:F2R1:PGT:PID:PS:SB            ./.:.:.:.:.:.:.:.:.:.  ...   \n",
       "\n",
       "      ./.:.:.:.:.:.:..36     ./.:.:.:.:.:.:..37     ./.:.:.:.:.:.:..38  \\\n",
       "0        ./.:.:.:.:.:.:.        ./.:.:.:.:.:.:.        ./.:.:.:.:.:.:.   \n",
       "1        ./.:.:.:.:.:.:.        ./.:.:.:.:.:.:.        ./.:.:.:.:.:.:.   \n",
       "2        ./.:.:.:.:.:.:.        ./.:.:.:.:.:.:.        ./.:.:.:.:.:.:.   \n",
       "3        ./.:.:.:.:.:.:.        ./.:.:.:.:.:.:.        ./.:.:.:.:.:.:.   \n",
       "4  ./.:.:.:.:.:.:.:.:.:.  ./.:.:.:.:.:.:.:.:.:.  ./.:.:.:.:.:.:.:.:.:.   \n",
       "\n",
       "      ./.:.:.:.:.:.:..39                                 ./.:.:.:.:.:.:..40  \\\n",
       "0        ./.:.:.:.:.:.:.                                    ./.:.:.:.:.:.:.   \n",
       "1        ./.:.:.:.:.:.:.                                    ./.:.:.:.:.:.:.   \n",
       "2        ./.:.:.:.:.:.:.                                    ./.:.:.:.:.:.:.   \n",
       "3        ./.:.:.:.:.:.:.                                    ./.:.:.:.:.:.:.   \n",
       "4  ./.:.:.:.:.:.:.:.:.:.  0|1:1,1:0.5:2:1,0:0,0:0|1:15751_A_T:15751:0,1,0,1   \n",
       "\n",
       "      ./.:.:.:.:.:.:..41     ./.:.:.:.:.:.:..42     ./.:.:.:.:.:.:..43  \\\n",
       "0        ./.:.:.:.:.:.:.        ./.:.:.:.:.:.:.        ./.:.:.:.:.:.:.   \n",
       "1        ./.:.:.:.:.:.:.        ./.:.:.:.:.:.:.        ./.:.:.:.:.:.:.   \n",
       "2        ./.:.:.:.:.:.:.        ./.:.:.:.:.:.:.        ./.:.:.:.:.:.:.   \n",
       "3        ./.:.:.:.:.:.:.        ./.:.:.:.:.:.:.        ./.:.:.:.:.:.:.   \n",
       "4  ./.:.:.:.:.:.:.:.:.:.  ./.:.:.:.:.:.:.:.:.:.  ./.:.:.:.:.:.:.:.:.:.   \n",
       "\n",
       "      ./.:.:.:.:.:.:..44 0/1:0,2:0.667:2:0,0:0,1:0,0,1,1  \n",
       "0        ./.:.:.:.:.:.:.                 ./.:.:.:.:.:.:.  \n",
       "1        ./.:.:.:.:.:.:.                 ./.:.:.:.:.:.:.  \n",
       "2        ./.:.:.:.:.:.:.                 ./.:.:.:.:.:.:.  \n",
       "3        ./.:.:.:.:.:.:.                 ./.:.:.:.:.:.:.  \n",
       "4  ./.:.:.:.:.:.:.:.:.:.           ./.:.:.:.:.:.:.:.:.:.  \n",
       "\n",
       "[5 rows x 56 columns]"
      ],
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>1</th>\n",
       "      <th>13273</th>\n",
       "      <th>.</th>\n",
       "      <th>G</th>\n",
       "      <th>C</th>\n",
       "      <th>..1</th>\n",
       "      <th>PASS</th>\n",
       "      <th>AS_FilterStatus=SITE;AS_SB_TABLE=0,0|0,0;ECNT=1;GERMQ=51;ROQ=54;DP=3;MBQ=0,20;MFRL=0,169;MMQ=60,47;MPOS=15;POPAF=7.3;TLOD=4.64</th>\n",
       "      <th>GT:AD:AF:DP:F1R2:F2R1:SB</th>\n",
       "      <th>./.:.:.:.:.:.:.</th>\n",
       "      <th>...</th>\n",
       "      <th>./.:.:.:.:.:.:..36</th>\n",
       "      <th>./.:.:.:.:.:.:..37</th>\n",
       "      <th>./.:.:.:.:.:.:..38</th>\n",
       "      <th>./.:.:.:.:.:.:..39</th>\n",
       "      <th>./.:.:.:.:.:.:..40</th>\n",
       "      <th>./.:.:.:.:.:.:..41</th>\n",
       "      <th>./.:.:.:.:.:.:..42</th>\n",
       "      <th>./.:.:.:.:.:.:..43</th>\n",
       "      <th>./.:.:.:.:.:.:..44</th>\n",
       "      <th>0/1:0,2:0.667:2:0,0:0,1:0,0,1,1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>13417</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>CGAGA</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td>AS_FilterStatus=SITE;AS_SB_TABLE=0,0|0,0;ECNT=...</td>\n",
       "      <td>GT:AD:AF:DP:F1R2:F2R1:SB</td>\n",
       "      <td>./.:.:.:.:.:.:.</td>\n",
       "      <td>...</td>\n",
       "      <td>./.:.:.:.:.:.:.</td>\n",
       "      <td>./.:.:.:.:.:.:.</td>\n",
       "      <td>./.:.:.:.:.:.:.</td>\n",
       "      <td>./.:.:.:.:.:.:.</td>\n",
       "      <td>./.:.:.:.:.:.:.</td>\n",
       "      <td>./.:.:.:.:.:.:.</td>\n",
       "      <td>./.:.:.:.:.:.:.</td>\n",
       "      <td>./.:.:.:.:.:.:.</td>\n",
       "      <td>./.:.:.:.:.:.:.</td>\n",
       "      <td>./.:.:.:.:.:.:.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>13868</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>map_qual</td>\n",
       "      <td>AS_FilterStatus=map_qual;AS_SB_TABLE=0,0|0,0;E...</td>\n",
       "      <td>GT:AD:AF:DP:F1R2:F2R1:SB</td>\n",
       "      <td>./.:.:.:.:.:.:.</td>\n",
       "      <td>...</td>\n",
       "      <td>./.:.:.:.:.:.:.</td>\n",
       "      <td>./.:.:.:.:.:.:.</td>\n",
       "      <td>./.:.:.:.:.:.:.</td>\n",
       "      <td>./.:.:.:.:.:.:.</td>\n",
       "      <td>./.:.:.:.:.:.:.</td>\n",
       "      <td>./.:.:.:.:.:.:.</td>\n",
       "      <td>./.:.:.:.:.:.:.</td>\n",
       "      <td>./.:.:.:.:.:.:.</td>\n",
       "      <td>./.:.:.:.:.:.:.</td>\n",
       "      <td>./.:.:.:.:.:.:.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>15118</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>map_qual</td>\n",
       "      <td>AS_FilterStatus=map_qual;AS_SB_TABLE=0,0|0,0;E...</td>\n",
       "      <td>GT:AD:AF:DP:F1R2:F2R1:SB</td>\n",
       "      <td>0/1:0,2:0.667:2:0,1:0,0:0,0,1,1</td>\n",
       "      <td>...</td>\n",
       "      <td>./.:.:.:.:.:.:.</td>\n",
       "      <td>./.:.:.:.:.:.:.</td>\n",
       "      <td>./.:.:.:.:.:.:.</td>\n",
       "      <td>./.:.:.:.:.:.:.</td>\n",
       "      <td>./.:.:.:.:.:.:.</td>\n",
       "      <td>./.:.:.:.:.:.:.</td>\n",
       "      <td>./.:.:.:.:.:.:.</td>\n",
       "      <td>./.:.:.:.:.:.:.</td>\n",
       "      <td>./.:.:.:.:.:.:.</td>\n",
       "      <td>./.:.:.:.:.:.:.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>15644</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>map_qual</td>\n",
       "      <td>AS_FilterStatus=map_qual;AS_SB_TABLE=0,0|0,0;E...</td>\n",
       "      <td>GT:AD:AF:DP:F1R2:F2R1:SB</td>\n",
       "      <td>./.:.:.:.:.:.:.</td>\n",
       "      <td>...</td>\n",
       "      <td>./.:.:.:.:.:.:.</td>\n",
       "      <td>./.:.:.:.:.:.:.</td>\n",
       "      <td>./.:.:.:.:.:.:.</td>\n",
       "      <td>./.:.:.:.:.:.:.</td>\n",
       "      <td>./.:.:.:.:.:.:.</td>\n",
       "      <td>./.:.:.:.:.:.:.</td>\n",
       "      <td>./.:.:.:.:.:.:.</td>\n",
       "      <td>./.:.:.:.:.:.:.</td>\n",
       "      <td>./.:.:.:.:.:.:.</td>\n",
       "      <td>./.:.:.:.:.:.:.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>15751</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>.</td>\n",
       "      <td>base_qual;haplotype</td>\n",
       "      <td>AS_FilterStatus=base_qual;AS_SB_TABLE=0,0|0,0;...</td>\n",
       "      <td>GT:AD:AF:DP:F1R2:F2R1:PGT:PID:PS:SB</td>\n",
       "      <td>./.:.:.:.:.:.:.:.:.:.</td>\n",
       "      <td>...</td>\n",
       "      <td>./.:.:.:.:.:.:.:.:.:.</td>\n",
       "      <td>./.:.:.:.:.:.:.:.:.:.</td>\n",
       "      <td>./.:.:.:.:.:.:.:.:.:.</td>\n",
       "      <td>./.:.:.:.:.:.:.:.:.:.</td>\n",
       "      <td>0|1:1,1:0.5:2:1,0:0,0:0|1:15751_A_T:15751:0,1,0,1</td>\n",
       "      <td>./.:.:.:.:.:.:.:.:.:.</td>\n",
       "      <td>./.:.:.:.:.:.:.:.:.:.</td>\n",
       "      <td>./.:.:.:.:.:.:.:.:.:.</td>\n",
       "      <td>./.:.:.:.:.:.:.:.:.:.</td>\n",
       "      <td>./.:.:.:.:.:.:.:.:.:.</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 56 columns</p>\n",
       "</div>"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 4
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": [
    "from utils import *\n",
    "\n",
    "NKI_submission = \"/Users/tushar/CancerModels/submission/NKI/IDC\"\n",
    "NKI_dir = \"/Users/tushar/CancerModels/pdxfinder-data/data/UPDOG/NKI\"\n",
    "template = \"/Users/tushar/CancerModels/pdxfinder-data/template/active_templates/\"\n",
    "\n",
    "\n",
    "def get_location_from_synonym(row):\n",
    "    symbol = row['symbol']\n",
    "    if isinstance(symbol, str):\n",
    "        pattern = \"(^|\\|)\" + symbol + \"($|\\|)\"\n",
    "        match = geneSymbol_location.loc[geneSymbol_location.Synonyms.str.contains(pattern)].reset_index(drop=True)\n",
    "        if len(match) > 1:\n",
    "            match = match.iloc[:1]\n",
    "        if len(match) > 0:\n",
    "            row['chromosome'], row['strand'], row['seq_start_position'], row['seq_end_position'], row['ncbi_gene_id'], \\\n",
    "            row['ensembl_gene_id'] = match['chromosome'][0], match['strand'][0], match['start'][0], match['end'][0], \\\n",
    "            match['GeneID'][0], match['ensembl_id'][0]\n",
    "    return row\n",
    "\n",
    "\n",
    "import requests\n",
    "import re\n",
    "\n",
    "\n",
    "def fetch_location_from_ucsc(url):\n",
    "    chr, location = None, None\n",
    "    try:\n",
    "        response = requests.get(url)\n",
    "        if response.ok:\n",
    "            response = response.json()\n",
    "            if \"positionMatches\" in response.keys() and 'matches' in response[\"positionMatches\"][\n",
    "                0].keys() and \"position\" in response[\"positionMatches\"][0]['matches'][0].keys():\n",
    "                response = response[\"positionMatches\"][0]['matches'][0]['position']\n",
    "                chr, location = response.split(':')\n",
    "                location = location.split('-')\n",
    "        return chr, location\n",
    "    except Exception as e:\n",
    "        print(f\"failed for symbol {url}\")\n",
    "        return None, None\n",
    "\n",
    "\n",
    "def get_location_from_ucsc_api(row):\n",
    "    if row['chromosome'] != '':\n",
    "        return row\n",
    "    url = f\"https://api.genome.ucsc.edu/search?search={row['symbol']}&genome=hg38\"\n",
    "    chr, location = fetch_location_from_ucsc(url)\n",
    "    if chr is not None and location is not None:\n",
    "        row['chromosome'] = chr.replace('chr', '')\n",
    "        row['seq_start_position'] = int(location[0])\n",
    "        row['seq_end_position'] = int(location[1])\n",
    "    if chr is None or location is None:\n",
    "        symbol = re.sub('\\.\\d*', '', row['symbol'])\n",
    "        url = f\"https://api.genome.ucsc.edu/search?search={symbol}&genome=hg38\"\n",
    "        chr, location = fetch_location_from_ucsc(url)\n",
    "        if chr is not None and location is not None:\n",
    "            row['chromosome'] = chr.replace('chr', '')\n",
    "            row['seq_start_position'] = int(location[0])\n",
    "            row['seq_end_position'] = int(location[1])\n",
    "    return row\n",
    "\n",
    "\n",
    "import numpy as np\n",
    "from tqdm import tqdm\n",
    "\n",
    "tqdm.pandas()\n",
    "\n",
    "\n",
    "# Define a function to convert log2 ratio to fold change\n",
    "def log2_to_fold_change(log2_ratio):\n",
    "    return 2 ** log2_ratio\n",
    "\n",
    "\n",
    "# Define a function to convert log2 ratio to log10 ratio\n",
    "def log2_to_log10(log2_ratio):\n",
    "    return log2_ratio / np.log10(2)\n",
    "\n",
    "\n",
    "genes = pd.read_csv('../resources/genes.tsv', sep='\\t')\n",
    "cna_matrix = read_metadata_with_fields(join(NKI_submission, 'cna/cna_cm_dcis_nki.txt'))\n",
    "cna_long_format = cna_matrix.melt(id_vars=[\"index\"], var_name=\"sample_id\", value_name=\"log2r_cna\").rename(\n",
    "    columns={\"index\": \"symbol\"})\n",
    "cna_long_format['log10r_cna'] = cna_long_format['log2r_cna'].apply(log2_to_log10)\n",
    "cna_long_format['fold_change'] = cna_long_format['log2r_cna'].apply(log2_to_fold_change)\n",
    "cna_long_format = cna_long_format.merge(\n",
    "    genes[['Symbol', 'chromosome', 'strand', 'start', 'end', 'GeneID', 'ensembl_id']], left_on='symbol',\n",
    "    right_on='Symbol', how='left').rename(\n",
    "    columns={'start': 'seq_start_position', 'end': 'seq_end_position', 'ensembl_id': 'ensembl_gene_id',\n",
    "             'GeneID': 'ncbi_gene_id'})\n",
    "#cna_long_format = cna_long_format.apply(get_location_from_synonym, axis=1)\n",
    "cna_long_format[cna_long_format['chromosome'].isna()].drop_duplicates('symbol')\n",
    "mapping = cna_long_format[['symbol', 'chromosome', 'strand', 'seq_start_position', 'seq_end_position', 'ncbi_gene_id',\n",
    "                           'ensembl_gene_id']].drop_duplicates().progress_apply(get_location_from_synonym, axis=1)\n",
    "other_mapping = mapping[mapping['chromosome'].isna()]\n",
    "other_mapping = other_mapping.fillna('').progress_apply(get_location_from_ucsc_api, axis=1)\n",
    "other_mapping = other_mapping.fillna('').progress_apply(get_location_from_ucsc_api, axis=1)\n",
    "other_mapping['strand'] = '+'\n",
    "mapping = pd.concat([other_mapping, mapping]).drop_duplicates(subset='symbol').reset_index(drop=True)\n",
    "# Perform an outer merge\n",
    "merged_df = pd.merge(cna_long_format, mapping, on=['symbol'], how='left', suffixes=('_df1', '_df2'))"
   ],
   "id": "initial_id"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": "",
   "id": "d621d64b83708760"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": "",
   "id": "80c8aebb2e84aea6"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": "",
   "id": "22517117e4405b8a"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": "",
   "id": "724460276075678b"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": "",
   "id": "47d70719b48fa108"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": [
    "cols = ['chromosome', 'strand', 'seq_start_position', 'seq_end_position', 'ncbi_gene_id', 'ensembl_gene_id']\n",
    "# Fill missing values in df1 columns with corresponding df2 columns\n",
    "for column in cols:\n",
    "    merged_df[column] = merged_df[f'{column}_df1'].combine_first(merged_df[f'{column}_df2'])\n",
    "\n",
    "# Drop the temporary columns used for merging\n",
    "merged_df.drop(columns=[f'{col}_df1' for col in cols], inplace=True)\n",
    "merged_df.drop(columns=[f'{col}_df2' for col in cols], inplace=True)\n",
    "merged_df.strand = merged_df.strand.astype(str).str.replace('-1.0', '-').str.replace('1.0', '+')\n",
    "merged_df.strand.unique()\n",
    "merged_df[merged_df['chromosome'] == ''].drop_duplicates(subset=['symbol'])\n",
    "merged_df[merged_df['chromosome'] == ''].shape[0] / merged_df.shape[0] * 100\n",
    "cna_long_format = merged_df[merged_df['chromosome'] != '']\n",
    "mapping[(mapping['chromosome'] != '') & (mapping['strand'] == '')]\n",
    "mapping\n"
   ],
   "id": "a1ff998ed6859418"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": [
    "submission_mms = read_metadata_with_fields(join(NKI_submission, 'molecular data/NKI_molecular_metadata_DCIS-sample.tsv'))\n",
    "submission_model_sample_dict = submission_mms[['model_id', 'sample_id']].drop_duplicates().groupby(['model_id'])['sample_id'].apply(set).to_dict()\n",
    "\n",
    "cna_template = read_metadata_with_fields(join(template, 'cna/cna_template-sheet.tsv'))\n",
    "cna_long_format = cna_long_format.merge(submission_mms[['sample_id', 'platform_id']], on='sample_id', how='left')\n",
    "for col in cna_template.columns:\n",
    "    if col not in cna_long_format.columns:\n",
    "        cna_long_format[col] = ''\n",
    "cna_long_format['platform_id'] = 'cna_' + cna_long_format['platform_id']\n",
    "for mid in submission_model_sample_dict.keys():\n",
    "    temp = cna_long_format[[sid in submission_model_sample_dict[mid] for sid in cna_long_format['sample_id']]]\n",
    "    temp[cna_template.columns].to_csv(join(NKI_submission, f'cna/NKI_cna_{mid}.tsv'), sep='\\t', index=False)"
   ],
   "id": "3da585a6053a0e1c"
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
